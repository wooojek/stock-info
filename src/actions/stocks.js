export const addStock = (stock) => ({
  type: 'ADD_STOCK',
  stock,
});

export const setStocks = (stocks) => ({
  type: 'SET_STOCKS',
  stocks,
});

export const startSetStocks = () => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const stocks = JSON.parse(localStorage.getItem('stocks')) || [];
      dispatch(setStocks(stocks));
      resolve();
    });
  };
}