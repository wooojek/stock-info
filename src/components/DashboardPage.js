import React from 'react';
import StockList from './StockList';

export const DashboardPage = () => (
  <div className='content-container'>
    <StockList />
  </div>
);

export default DashboardPage;