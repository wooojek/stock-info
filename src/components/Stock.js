import React from 'react';
import numeral from 'numeral';

const { ALPHAVANTAGE_API_KEY } = process.env;

export class Stock extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.name,
      symbol: props.symbol,
      market: props.market,
      domain: props.domain || '',
      logo: props.logo || 'https://via.placeholder.com/64',
    }
  }

  getImage = () => {
    const query = this.state.name.split(' ')[0];
    return fetch(`https://autocomplete.clearbit.com/v1/companies/suggest?query=${query}`)
      .then((response) => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(([{ domain, logo }]) => {
        return {
          domain,
          logo,
        };
      })
      .catch((e) => console.log(e));
  }

  getStockInfo = () => {
    return fetch(`https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=${this.state.symbol}&apikey=${ALPHAVANTAGE_API_KEY}`)
        .then((response) => {
          if (!response.ok) {
            throw Error(response.statusText);
          }
          return response.json();
        })
        .then((data) => {
          if (data['Note']) {
            throw Error(data['Note']);
          }
          const stockData = data['Global Quote'];
          return {
            price: stockData['05. price'],
            closed: stockData['07. latest trading day'],
            change: stockData['09. change'],
            changePercent: stockData['10. change percent'],
          };
        })
        .catch((e) => console.log(e));
  }

  async componentDidMount() {
    try {
      const { domain, logo } = await this.getImage();
      const details = await this.getStockInfo();

      this.setState(() => ({
        details,
        domain,
        logo,
      }));
    } catch (e) {
      console.log(e);
    }
  }

  render() {
    const details = this.state.details;
    const market = this.state.market;
    return (
      <div className="list-item">
        <div className="stock-container">
          <div
            className="stock-img"
            style={{ backgroundImage: `url(${this.state.logo})` }}
          >
          </div>
          <div className="stock-info">
            <div className="stock-info__names">
              <span className="bold">{this.state.name}</span>
              <span>{this.state.symbol}</span>
              {this.state.domain ? (<span>{this.state.domain}</span>) : false}
            </div>
            {market ? (
              <div className="stock-info__market">
                <span>{market.region}</span>
                <span>{market.marketOpen} - {market.marketClose}</span>
                <span>{market.timezone}</span>
              </div>
            ) : (<div></div>)}
            {details ? (
              <div className="stock-info__details">
                <span className="bold">{numeral(details.price).format('$0,0.00')}</span>
                <span className={details.change >= 0 ? 'stock--up' : 'stock--down'}>
                  {numeral(details.change).format('0.00')} ({numeral(details.changePercent.slice(0, -2)).format('0.00')}%)
                </span>
                <span>Closed: {details.closed}</span>
              </div>
            ) : (<div></div>)}

          </div>
        </div>
      </div>
    )
  }
}

export default Stock;