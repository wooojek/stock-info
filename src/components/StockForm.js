import React from 'react';

export default class StockForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.stock ? props.stock.name : '',
      error: '',
    }
  }

  onNameChange = (e) => {
    const name = e.target.value.toUpperCase();
    this.setState(() => ({ name }))
  }

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.name) {
      this.setState(() => ({ errorState: 'Please provide name.' }));
    } else {
      this.setState(() => ({ errorState: '' }));
      this.props.onSubmit({
        name: this.state.name,
      });
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.errorState && (<p className="form__error">{this.state.errorState}</p>)}
        <input
          type="text"
          className="text-input"
          placeholder="Company symbol"
          autoFocus
          value={this.state.name}
          onChange={this.onNameChange}
        />
        <div>
          <button className="button">Track</button>
        </div>
      </form>
    )
  }
};