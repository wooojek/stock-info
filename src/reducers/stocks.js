const defaultStockState = [
  {
    name: 'Apple',
  },
  {
    name: 'Google',
  },
];

export default (state = defaultStockState, action) => {
  switch (action.type) {
    case 'ADD_STOCK':
      return [
        action.stock,
        ...state,
      ];
    default:
      return state;
  }
};