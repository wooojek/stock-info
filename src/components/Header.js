import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { logout } from '../actions/auth';

export const Header = ({ logout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/">
          <h1>Stock Info</h1>
        </Link>
        <Link className="header__element--link" to="/add">
          <h3>Add new company</h3>
        </Link>
        <button className="button button--link header__logout" onClick={logout}>Logout</button>
      </div>

    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  logout: () => dispatch(logout()),
});

export default connect(undefined, mapDispatchToProps)(Header);