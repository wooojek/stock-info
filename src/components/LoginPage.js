import React from 'react';
import { connect } from 'react-redux';
import uuid from 'uuid/v1';
import { login } from '../actions/auth';

const uid = uuid();

export const LoginPage = ({ login }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Stock info</h1>
      <p>Be up to date with your stocks!</p>
      <button className="button" onClick={login}>Login</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  login: () => dispatch(login(uid)),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);