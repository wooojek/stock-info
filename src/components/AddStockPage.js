import React from 'react';
import { connect } from 'react-redux';
import StockForm from './StockForm';
import { startAddStock } from '../actions/stocks';

export class AddStockPage extends React.Component {
  onSubmit = (stock) => {
    this.props.startAddStock(stock);
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Add Stock</h2>
          </div>
        </div>
        <div className="content-container">
          <StockForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
};

const mapDispatchToProps = (dispatch) => ({
  startAddStock: (stock) => dispatch(startAddStock(stock)),
});

export default connect(undefined, mapDispatchToProps)(AddStockPage);