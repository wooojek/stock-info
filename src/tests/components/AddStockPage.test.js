import React from 'react';
import { shallow } from 'enzyme';
import { AddStockPage } from '../../components/AddStockPage';
import stocks from '../fixtures/stocks';

let startAddStock, history, wrapper;

beforeEach(() => {
  startAddStock = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddStockPage startAddStock={startAddStock} history={history} />);
});

test('should render AddStockPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('StockForm').prop('onSubmit')(stocks[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startAddStock).toHaveBeenLastCalledWith(stocks[1]);
});