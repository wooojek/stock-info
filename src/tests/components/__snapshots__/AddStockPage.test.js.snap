// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddStockPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        Add Stock
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <StockForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
